tests/ttnn/unit_tests/gtests/test_multiprod_queue.cpp

```diff
@@ -90,3 +91,61 @@ TEST_F(MultiCommandQueueSingleDeviceFixture, TestMultiProducerLockBasedQueue) {
     t1.join();
 
 }
+
+TEST_F(MultiCommandQueueSingleDeviceFixture, TestMultiAppThreadSync) {
+    // Have 2 threads use the same HW CQ and use host <-> device event synchronization
+    // to guarantee ordering
+    Device* device = this->device_;
+    // Enable async engine and set queue setting to lock_based
+    device->set_worker_mode(WorkExecutorMode::ASYNCHRONOUS);
+    device->set_worker_queue_mode(WorkerQueueMode::LOCKBASED);
+
+    MemoryConfig mem_cfg = MemoryConfig{
+        .memory_layout = tt::tt_metal::TensorMemoryLayout::INTERLEAVED,
+        .buffer_type = BufferType::DRAM,
+        .shard_spec = std::nullopt
+    };
+    uint32_t write_cq = 0;
+    uint32_t read_cq = 0;
+    uint32_t tensor_buf_size = 1024 * 1024;
+    uint32_t datum_size_bytes = 2;
+
+    std::shared_ptr<Event> write_event = std::make_shared<Event>();
+    std::shared_ptr<Event> read_event = std::make_shared<Event>();
+
+    ttnn::Shape tensor_shape = ttnn::Shape(Shape({1, 1, 1024, 1024}));
+    auto host_data = std::shared_ptr<bfloat16 []>(new bfloat16[tensor_buf_size]);
+    auto allocated_buffer = ttnn::allocate_buffer_on_device(tensor_buf_size * datum_size_bytes, device, tensor_shape, DataType::BFLOAT16, Layout::TILE, mem_cfg);
+    auto allocated_storage = tt::tt_metal::DeviceStorage{allocated_buffer};
+    auto allocated_tensor = Tensor(allocated_storage, tensor_shape, DataType::BFLOAT16, Layout::TILE);
+    auto readback_data = std::shared_ptr<bfloat16 []>(new bfloat16[tensor_buf_size]);
+
+    std::thread t0([&] () {
+        for (int j = 0; j < 1000; j++) {
+            if (j != 0) {
+                ttnn::event_synchronize(read_event);
+            }
+            read_event = std::make_shared<Event>();
+            for (int i = 0; i < tensor_buf_size; i++) {
+                host_data[i] = bfloat16(static_cast<float>(2 + j));
+            }
+            ttnn::write_buffer(write_cq, allocated_tensor, {host_data});
+            ttnn::record_event(device->command_queue(write_cq), write_event);
+        }
+    });
+
+    std::thread t1([&] () {
+        for (int j = 0; j < 1000; j++) {
+            ttnn::event_synchronize(write_event);
+            write_event = std::make_shared<Event>();
+            ttnn::read_buffer(read_cq, allocated_tensor, {readback_data});
+            for (int i = 0; i < tensor_buf_size; i++) {
+                EXPECT_EQ(readback_data[i], host_data[i]);
+            }
+            ttnn::record_event(device->command_queue(read_cq), read_event);
+        }
+    });
```

Comment:
Can you help to understand what exactly do you test here?

Reply:
maybe I am confused because event names are not clear

Reply:
This test verifies that the `event_synchronize`API, which is meant to stall the calling thread works as expected.
Thread 0 is the writer. Thread 1 is the reader, i.e. for iteration `i` thread 1 cannot read until thread 0 has written.

Hence, we stall thread 1 until the write is completed using `ttnn::event_synchronize(write_event)`.
Similarly, thread 0 cannot write to a memory location until thread 1 has read the data at that memory location. `ttnn::event_synchronize(read_event)` is used to stall the writer until the reader notifies completion.

This should be a similar setup to what Moreh is using in their multi-threaded runtime with multiple threads using a single command queue and using events for synchronization.

This test should have been here from the beginning, since it regresses on one of the main use cases of event synchronization.

Reply:
We should embed most of what you've described here into the code; it's helpful context!

Reply:
I've updated the test with a description, and the purpose of using the event API